Adds Alloy Materials

TemporalOroboros opened the pull request:
Adds and implements alloy materials

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Takes several materials that were mostly fluff and converts them into actual usable materials.

Messes with material code a bit to make alloys recycle back into their component materials.

Adds the alloy materials to their in-game stacks.

Materials added:
- Plasteel
- Plastitanium
- Plasmaglass
- Titaniumglass
- Plastitanium Glass
- Alien Alloy

These should have a once-over from someone who actually gets balance as I have no sense for it and just picked numbers that felt good.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Makes plasteel/plastitanium/plasmaglass and the rest able to have separate properties from their component materials. It doesn't make much sense that the materials used to seal off the supermatter chamber from the rest of the station would be prone to exploding when heated.

Allows for further expansion of materials, possibly including actual functional metallurgy and smelting at some point in the _very_ distant future.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Makes plasteel/plastitanium and a few other alloys actual materials.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros committed:
Alloys

Adds and implements alloy materials

TemporalOroboros commented:
Oh right, I should probably show how these look.

![materials](https://user-images.githubusercontent.com/53582584/92848671-559e2a80-f39f-11ea-818d-5d57eafc918e.jpg)
![materials2 2](https://user-images.githubusercontent.com/53582584/92848685-58991b00-f39f-11ea-87e0-4e27fc883beb.jpg)

Livrah commented:
Does custom toolboxes has sprites too?

TemporalOroboros commented:
You're right I forgot about those, but custom toolboxes just inherit the color of the materials they are made from. Basically thing a greyscale toolbox with the same color and opacity as the rest of the objects in the pictures and you should have a decent idea of what they look like.

necromanceranne commented:
I was saying I'd do this eventually but I got sidetracked hard. Glad someone is doing it, this always bugged me!

ATH1909 commented:
The modifiers (strength, integrity, and armor) for these alloys are far superior to those of most normal materials (even runite, except in bullet armor, but knight armor will have a dogshit bullet armor value regardless of which material you use, so it doesn't really matter). Cleric maces and knight armor are pretty underpowered in general right now, though, so I'd suggest bringing those modifiers back in line but buffing the base stats of cleric maces and knight armor (or at least making knight armor more well-rounded instead of only being good for melee armor).

ATH1909 commented on code/datums/materials/alloys.dm:
```diff
@@ -0,0 +1,124 @@
+	desc = "The heavy duty result of infusing iron with plasma."
+	color = "#828282"
+	value_per_unit = 0.135
+	strength_modifier = 1.5
```
force 21 mace that can break down normal doors go brrrrrrr

(tbh, a force 21 weapon isn't too out of line with other space loot, but the problem here is that cleric maces themselves are too weak, not that other materials are too weak, and plasteel eclipses those other materials in power level by a significant amount. nerf these alloys so that we can buff cleric maces and knight armor.)

TemporalOroboros replied:
Plasteel's strangth modifier has been to parity with titanium. Is this sufficient?

ATH1909 commented on code/datums/materials/alloys.dm:
```diff
@@ -0,0 +1,124 @@
+	value_per_unit = 0.135
+	strength_modifier = 1.5
+	integrity_modifier = 1.8 // Heavy duty.
+	armor_modifiers = list(MELEE = 1.5, BULLET = 1.5, LASER = 1.3, ENERGY = 1.3, BOMB = 2, BIO = 1, RAD = 1.8, FIRE = 1.5, ACID = 1.5)
```
knight suit with 75 melee armor go brrrrrrrrr

TemporalOroboros replied:
It's been nerfed... slightly. Plasteel also now add slowdown to any items it's included in. Is this sufficient?

TemporalOroboros replied:
I just checked and it turns out plastic has a 50% boost to melee armor for some reason. It just feels wrong for plasteel to have less melee armor than plastic.

Qustinnus replied:
the hell? I'd say nerf plastic,

ATH1909 commented on code/datums/materials/alloys.dm:
```diff
@@ -0,0 +1,124 @@
+	name = "alien alloy"
+	desc = "An extremely dense alloy similar to plasteel in composition. It requires exotic metallurgical processes to create."
+	color = "#313a68"
+	strength_modifier = 1.8 // It's twice the density of plasteel and just as durable. Getting hit with it is going to HURT.
```
an alien alloy cleric mace (a one-handed weapon) would have a base force of 25, greater than that of a two-handed fire axe. again, that's probably okay, seeing as how you'd have to raid a specific space ruin AND get alien alloy to make such a mace, but I figured that I should point that out.

TemporalOroboros replied:
I've nerfed this to parity with adamantine, it should be a bit more reasonable now.

Fikou commented on code/datums/materials/alloys.dm:
```diff
@@ -0,0 +1,124 @@
+	desc = "The heavy duty result of infusing iron with plasma."
+	color = "#828282"
+	value_per_unit = 0.135
+	strength_modifier = 1.5
```
bro this is what i made adamantine strength modifier as the highest one and now you can get it with a roundstart sheet

TemporalOroboros replied:
I've nerfed this to about parity with titanium That should be enough.

Fikou commented on code/datums/materials/alloys.dm:
```diff
@@ -0,0 +1,124 @@
+	value_per_unit = 0.135
+	strength_modifier = 1.5
+	integrity_modifier = 1.8 // Heavy duty.
+	armor_modifiers = list(MELEE = 1.5, BULLET = 1.5, LASER = 1.3, ENERGY = 1.3, BOMB = 2, BIO = 1, RAD = 1.8, FIRE = 1.5, ACID = 1.5)
```
this is almost like mythril stats

TemporalOroboros replied:
Nerfed to be about parity with adamantine. It should be a bit more reasonable now.

Qustinnus commented on code/__DEFINES/materials.dm:
```diff
@@ -7,6 +7,22 @@
 
+// Breakdown flags used when decomposing alloys.
+/// Can reduce an alloy into its component materials.
+#define BREAKDOWN_ALLOYS				(1<<0)
```
Is the idea of this that right now they're all the same but later on we can add more options? Seeing as they all have the same value

TemporalOroboros replied:
Pretty much. At some point there should be different methods for breaking down different alloys but at the moment couldn't be bothered to replace the entire recycling system.

Qustinnus replied:
Sounds fine to me, maybe add a comment that says this to clear up any confusion?

TemporalOroboros replied:
Comment added.

Qustinnus commented on code/datums/materials/alloys.dm:
```diff
@@ -0,0 +1,124 @@
+  *
+  * An alloy of iron and plasma.
+  */
+/datum/material/alloy/plasteel
```
As mentioned in other comments, a lot of these materials are FAR too strong in the current material balance, Value, strength and armor ratings need to go down by a lot, Most materials should not be a perfect solution to everything, and instead have buffs and debuffs. If you want plasteel to for example be as strong as it is, wearing plasteel objects should slow you down etc. Feel free to add new components/elements for that (Optionally in a separate PR). However, if things like that aren't added these stats are simple too high, we can talk about this more in-depth on Discord if you'd like.

TemporalOroboros replied:
I've nerfed them a bit, they are probably still way too strong though.

Qustinnus replied:
I think so too, unless making alloys becomes a more difficult process I don't think they should be much more powerful than the materials theyre made of.

TemporalOroboros replied:
Metallurgy can wait for another time. Any advice for what else to nerf?

Qustinnus commented on code/datums/materials/alloys.dm:
```diff
@@ -0,0 +1,124 @@
+/** Materials made from other materials.
+  */
+/datum/material/alloy
```
This type is now included in any places that does subtypesof(), I don't think thats a lot of places but I do know SSmaterials will initialize this material now, can you make sure that that doesn't happen? :)

TemporalOroboros replied:
Should have been fixed, though you should probably give it another look.

Qustinnus reviewed (commented):
This looks really high-quality overall, and I think it's an awesome project. I've been interested in adding this myself but been burnt out on mats lately, so I'm glad you've taken an interest. The only real worry is the statistics but I'm sure we can come to a solution for that.

TemporalOroboros committed:
Merge branch 'master' into alloys

TemporalOroboros committed:
Nerf

Nerfs the armor, strength, and integrity values of the alloy materials. Gives plasteel and alien alloy slowdown. Gives alien alloy regen.

TemporalOroboros committed:
Nerf2

Further reduces armor values. Brings alien alloy to near-parity with adamantine.

TemporalOroboros committed:
Comments and more nerfs.

Adds some comments and descriptions. Nerfs plasteel, plastitanium, and alien alloy.

TemporalOroboros committed:
/datum/element/alloy_regen/var/id_arg_index

TemporalOroboros committed:
Slight plasteel nerf.

Nerfs plasteel's strength mod a bit and brings plasteel's melee armor mod to pairity with titanium

TemporalOroboros committed:
Un-nerf plasteel armor.

trerri commented:
Titanium (glass) should look white, and plastitanium should have something to distinguish it from regular iron (probably outside of the scope but eh).
Alien alloy looks too dark, they kind of all look very dark

Qustinnus commented:
the glass should probably also be transparent

TemporalOroboros commented:
I basically just copied the color and transparency values for the sheets. I'll change them.

TemporalOroboros committed:
Merge branch 'master' into alloys

TemporalOroboros commented:
I've made some changes to both color and alpha. Is it any better? I can't tell.
![Materials3](https://user-images.githubusercontent.com/53582584/93027972-8364b880-f5c5-11ea-8038-fc1351e371ef.PNG)

TemporalOroboros committed:
Changes color and alpha values for the alloys.

Requested by Qustinnus & trerri

TemporalOroboros commented:
@Qustinnus Is this any better?

Qustinnus commented:
yeah looks great imo

TemporalOroboros committed:
Merge branch 'master' into alloys

Qustinnus commented on code/datums/materials/alloys.dm:
```diff
@@ -0,0 +1,163 @@
+	if(!isitem(target_item))
+		return
+
+	target_item.slowdown = initial(target_item.slowdown)
```
This can cause edge-cases when multiple sources affect slowdown.

TemporalOroboros replied:
I'm not sure how to fix this without implementing a slowdown modifier system for items. Do you have any advice because that seems a bit out of scope?

Qustinnus replied:
Could you instead just remove the slowdown again? ofcourse also not ideal but I think it has less edgecases.

TemporalOroboros replied:
I've removed the slowdown modifier. It's a significant buff which is the *opposite* of what we're trying to do right now, I'll try to think of something else.

Qustinnus replied:
I think you misunderstood, I meant that you do target_item.slowdown -= (whatever calcs you did to add it)

Qustinnus replied:
instead of setting it to initial

TemporalOroboros replied:
I'll do that.

TemporalOroboros replied:
Should be done, I had to add proper amount handling to all of the material removal procs. it looks like bananium and possibly hot ice already had it. \<sarcasm>I'm sure that wouldn't have broken all if anything ever bothered to remove materials.\</sarcasm>

Qustinnus commented on code/datums/materials/_material.dm:
```diff
@@ -12,6 +12,8 @@ Simple datum which is instanced once per type and is used for every object of sa
 	///Base alpha of the material, is used for greyscale icons.
 	var/alpha
+	///Bitflags that influence how SSmaterials handles this material.
+	var/material_flags = MATERIAL_INIT_MAPLOAD
```
We now have a similarly named var on atoms which handles how the atom treats materials, I can imagine that can confuse people.

TemporalOroboros replied:
This one I can solve easily enough.

Qustinnus replied:
<3

TemporalOroboros replied:
Is `init_flags` a good enough name?

Qustinnus replied:
yeah seems fine :D We can always change it if we want to add different types of flags :)

TemporalOroboros committed:
Merge branch 'master' into alloys

TemporalOroboros committed:
material_flags -> init_flags

TemporalOroboros committed:
removes material slowdown

Removes slowdown from plasteel and alien alloy.
TODO: Add a proper slowdown modifier system for items.

TemporalOroboros committed:
Revert "removes material slowdown"

This reverts commit 3df8555aa7fa89997776a84d1fe4d6ae7c6ea9d6.

TemporalOroboros committed:
properly subtract material slowdown

Makes plasteel and alien alloy properly remove their slowdown.

Adds amount handling to all material removal.

TemporalOroboros committed:
Merge branch 'master' into alloys

TemporalOroboros committed:
Merge branch 'master' into alloys

Qustinnus reviewed (approved):
This PR is good to merge btw

TheBonded commented:
Does this mean we can make plastanium toolboxes in autolathes? And if so, what do all the different alloys have different from iron toolboxes?

Twaticus reviewed (approved):
more walls for the 3/4th wall gods

TemporalOroboros commented:
@TheBonded Yes, you can make plastitanium toolboxes. Right now most of the materials just have varying durability/strength. Plasteel and Alien Alloy have some slowdown mostly to compensate for their high armor values. Anything made of alien alloy also regenerates slowly.

TemporalOroboros commented:
@Twaticus The materials use the greyscale wall sprites so it shouldn't be significantly more work for you.

Qustinnus commented:
the entire point of mat datums was less wall sprites if anything

Fikou commented:
does this fix https://github.com/tgstation/tgstation/issues/54036

Qustinnus commented:
yes most likely,

TemporalOroboros commented:
@Fikou Sort of? It doesn't make it possible to make the alien ship walls, but it does _technically_ make it possible to make alien alloy walls, just the generic material version.

TemporalOroboros committed:
Merge branch 'master' into alloys

TemporalOroboros committed:
Merge branch 'master' into alloys

TemporalOroboros committed:
Merge branch 'master' into alloys

msgerbs commented:
What's this waiting on?

Qustinnus commented:
A maintainer that cares about material datums

LemonInTheDark commented on code/datums/components/material_container.dm:
```diff
@@ -73,17 +76,17 @@
 
 /// Proc used for when player inserts materials
-/datum/component/material_container/proc/user_insert(obj/item/I, mob/living/user)
+/datum/component/material_container/proc/user_insert(obj/item/I, mob/living/user, breakdown_flags=src.breakdown_flags)
```
Why the src.breakdown_flags? is that needed to achieve this?

TemporalOroboros replied:
I put that there so that if this is called from elsewhere, such as the remote materials component, and this isn't overridden it will default to breaking down the materials through its own methods. I could move it into the proc body if you want me to.

LemonInTheDark commented on code/datums/elements/alloy_regen.dm:
```diff
@@ -0,0 +1,84 @@
+/** Object integrity regeneration element added by alien alloy.
+  */
+/datum/element/alloy_regen
```
Is there any reason this element has to be alloy specific?

TemporalOroboros replied:
Probably not, I'll rename it.

TemporalOroboros replied:
Done.

LemonInTheDark commented on code/game/objects/obj_defense.dm:
```diff
@@ -11,6 +11,9 @@
 	damage_amount = run_obj_armor(damage_amount, damage_type, damage_flag, attack_dir, armour_penetration)
 	if(damage_amount < DAMAGE_PRECISION)
 		return
+	if(SEND_SIGNAL(src, COMSIG_OBJ_TAKE_DAMAGE, damage_amount, damage_type, damage_flag, sound_effect, attack_dir, armour_penetration) & COMPONENT_NO_TAKE_DAMAGE)
+		return
+
```
Why put this here?

TemporalOroboros replied:
Because this is after all of the checks and damage mitigation. At this point the object is guaranteed to actually take damage.

LemonInTheDark reviewed (commented):
Some "I'm 60 what does this do", on the whole this looks good, if the numbers aren't sane I'm going to nerf floyd out of existence

TemporalOroboros committed:
alloy_regen->obj_regen

Renames the alloy regen element.

TemporalOroboros committed:
Merge branch 'master' into alloys
